### ISSUE TITLE
confignetwork postscript tries to invoke non-existing nmcli when configuring bond during deployment

### ISSUE THREAD

**Issue by kcgthb** [xcat2/xcat-core#5420]
The `confignetwork` postscript tries to bring bonding interfaces up and down during deployment, and fails because `nmcli` doesn't exist in the chroot'ed deployment environment.

Here's an excerpt of `/var/log/xcat/xcat.log` in debug mode while configuring and `bond0` interface. The node `nics` definition is:
```
Object name: sh-dtn01
    nicaliases.bond0=sh-dtn01
    nicaliases.bmc=sh-dtn01
    nicaliases.ib0=sh-dtn01
    nicdevices.bond0=em2|em3
    nicextraparams.em1=MTU=9000
    nicextraparams.bond0=MTU=9000
    nicextraparams.ib0=MTU=65520 CONNECTED_MODE=yes
    nicips.bond0=171.66.97.121
    nicips.bmc=10.11.0.81
    nicips.ib0=10.9.0.81
    nicnetworks.bond0=net_stanford
    nicnetworks.bmc=net_ipmi
    nicnetworks.ib0=net_ib_100
    nictypes.em2=ethernet
    nictypes.em3=ethernet
    nictypes.bond0=bond
    nictypes.bmc=bmc
    nictypes.ib0=infiniband
```

```
[...]
[I]: >> TYPE="Bond"
+ return 0
+ read __msg
+ log_info '>> BONDING_OPTS="mode=802.3ad miimon=100"'
+ local '__msg=>> BONDING_OPTS="mode=802.3ad miimon=100"'
+ log_print_default '[I]: >> BONDING_OPTS="mode=802.3ad miimon=100"'
+ printf '%s\n' '[I]: >> BONDING_OPTS="mode=802.3ad miimon=100"'
[I]: >> BONDING_OPTS="mode=802.3ad miimon=100"
+ return 0
+ read __msg
+ log_info '>> DHCLIENTARGS="-timeout 200"'
+ local '__msg=>> DHCLIENTARGS="-timeout 200"'
+ log_print_default '[I]: >> DHCLIENTARGS="-timeout 200"'
+ printf '%s\n' '[I]: >> DHCLIENTARGS="-timeout 200"'
[I]: >> DHCLIENTARGS="-timeout 200"
+ return 0
+ read __msg
+ return 0
+ '[' xnet_stanford '!=' x ']'
++ ifdown bond0
++ ifup bond0
+ lines='/sbin/ifup: line 73: nmcli: command not found
WARN      : [/etc/sysconfig/network-scripts/ifup-eth] Unable to start slave device ifcfg-em2 for master bond0.
/sbin/ifup: line 73: nmcli: command not found
WARN      : [/etc/sysconfig/network-scripts/ifup-eth] Unable to start slave device ifcfg-em3 for master bond0.'
+ rc=0
+ '[' 0 -ne 0 ']'
+ wait_for_ifstate bond0 UP 200 1
+ local ifname=bond0
+ local ifstate=UP
+ local tryCnt=200
+ local tryInt=1
+ local state=
+ local i
+ (( i=0 ))
+ '[' 0 -lt 200 ']'
++ ip link show bond0
+ lines='8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
+ echo '8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
+ grep -sq 'state UP'
+ rc=1
+ '[' 1 -eq 0 ']'
+ '[' 1 -ne 0 ']'
++ echo '8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
++ grep state
++ sed -e 's/^.*state \([a-zA-Z]\+\) .*$/\1/g'
+ state=DOWN
+ log_info 'State of "bond0" was "DOWN" instead of expected "UP". Wait 0 of 200 with interval 1.'
+ local '__msg=State of "bond0" was "DOWN" instead of expected "UP". Wait 0 of 200 with interval 1.'
+ log_print_default '[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 0 of 200 with interval 1.'
+ printf '%s\n' '[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 0 of 200 with interval 1.'
[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 0 of 200 with interval 1.
+ return 0
+ sleep 1
+ (( i+=1 ))
+ '[' 1 -lt 200 ']'
++ ip link show bond0
+ lines='8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
+ echo '8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
+ grep -sq 'state UP'
+ rc=1
+ '[' 1 -eq 0 ']'
+ '[' 1 -ne 0 ']'
++ echo '8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
++ grep state
++ sed -e 's/^.*state \([a-zA-Z]\+\) .*$/\1/g'
+ state=DOWN
+ log_info 'State of "bond0" was "DOWN" instead of expected "UP". Wait 1 of 200 with interval 1.'
+ local '__msg=State of "bond0" was "DOWN" instead of expected "UP". Wait 1 of 200 with interval 1.'
+ log_print_default '[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 1 of 200 with interval 1.'
+ printf '%s\n' '[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 1 of 200 with interval 1.'
[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 1 of 200 with interval 1.
+ return 0
+ sleep 1
+ (( i+=1 ))
+ '[' 2 -lt 200 ']'
++ ip link show bond0
+ lines='8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
+ echo '8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
+ grep -sq 'state UP'
+ rc=1
+ '[' 1 -eq 0 ']'
+ '[' 1 -ne 0 ']'
++ echo '8: bond0: <NO-CARRIER,BROADCAST,MULTICAST,MASTER,UP> mtu 9000 qdisc noqueue state DOWN mode DEFAULT group default qlen 1000
++ grep state
    link/ether 8e:34:d1:25:5f:6a brd ff:ff:ff:ff:ff:ff'
++ sed -e 's/^.*state \([a-zA-Z]\+\) .*$/\1/g'
+ state=DOWN
+ log_info 'State of "bond0" was "DOWN" instead of expected "UP". Wait 2 of 200 with interval 1.'
+ local '__msg=State of "bond0" was "DOWN" instead of expected "UP". Wait 2 of 200 with interval 1.'
+ log_print_default '[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 2 of 200 with interval 1.'
+ printf '%s\n' '[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 2 of 200 with interval 1.'
[I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 2 of 200 with interval 1.
[...]
```

That seem to happen in `nicutils.sh`, in `create_bond_interface()`
https://github.com/xcat2/xcat-core/blob/46c1bf5c5ef953bc7776f4c035e92199cef34612/xCAT/postscripts/nicutils.sh#L1489

There are many functions in `nicutils.sh`, which is sourced by `confignetwork`, which use `ifup`/`ifdown` and those will fail when deploying RHEL/CentOS diskful systems, since `nmcli` doesn't exist in the deployment environment.

This seem to have been mentioned in 
https://github.com/xcat2/xcat-core/issues/4754#issuecomment-363026144:
> when the scripts `confignetworks` and `confignics` are running in the installer(chroot environment), they should only generate the necessary configuration files , should not invoke operations like `ifdown/ifup`

But apparently, there are still cases where `nmcli` is used during deployment.
What would you suggest to avoid this?

**Comment by bybai** [comment]
Hi @kcgthb , what is your OS?

**Comment by bybai** [comment]
@kcgthb, could you try to use  following to replace the " line 1489: lines=`$ifdown $ifname; $ifup $ifname 2>&1`" in your environment?
```
lines=`ip link set $ifname down; ip link set $ifname up 2>&1`
```

**Comment by kcgthb** [comment]
Hi @bybai 
We're deploying CentOS 7 diskful images.
And yes, using `ip link` instead of `ifup`/`ifdown` seems a much better approach: error messages about `nmcli` don't appear anymore.

**Comment by bybai** [comment]
@kcgthb , after using "ip link", Does the bond "state UP"?

**Comment by kcgthb** [comment]
@bybai Yes, it takes a few tries, but it succeeds:
```
Jul 23 16:54:09 sh-dtn01 xcatpostinit1: [I]: ip link set bond0 up
Jul 23 16:54:09 sh-dtn01 xcatpostinit1: [I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 0 of 200 with interval 1.
Jul 23 16:54:10 sh-dtn01 xcatpostinit1: [I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 1 of 200 with interval 1.
Jul 23 16:54:11 sh-dtn01 xcatpostinit1: [I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 2 of 200 with interval 1.
Jul 23 16:54:12 sh-dtn01 xcatpostinit1: [I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 3 of 200 with interval 1.
Jul 23 16:54:13 sh-dtn01 xcatpostinit1: [I]: State of "bond0" was "DOWN" instead of expected "UP". Wait 4 of 200 with interval 1.
Jul 23 16:54:14 sh-dtn01 xcatpostinit1: [I]: [ip.link] >> 8: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 9000 qdisc noqueue state UP mode DEFAULT group default qlen 1000
Jul 23 16:54:14 sh-dtn01 xcatpostinit1: [I]: [ip.link] >>     link/ether 24:6e:96:98:bd:b1 brd ff:ff:ff:ff:ff:ff
```

**Comment by bybai** [comment]
Hi @kcgthb, 
I have questions need your confirm:
1. Do you run `confignetwork` in postscripts or postbootscripts? 

2. Why there is no nmcli in your chroot'ed deployment environment? Do you delete NetworkManager in your env?
I think NetworkManager should be installed in Centos7 by default. 

3. ``ip link set <dev> down`` and ``ip link set <dev> up`` can not apply ifcfg-<dev> configure file, although it can works in your env, but it is not a good solution.

**Comment by kcgthb** [comment]
@bybai 
We're running the `confignetwork` script as a `postscript`, not as a `postbootscript`, so it's executed in the anaconda environment, before the OS is fully deployed. That wouldn't be an issue if `confignetworks` were run as a `postbootscript`, since it would be executed once the OS is deployed and NetworkManager would be installed, indeed.

1. there is no NetworkManager nor `nmcli` in the kickstart/anaconda environment used by RedHat/CentOS 7 to install the OS. I don't delete it, it's just not there. 

2. it's not really a question :) but maybe the `confignetwork` script could just check in what context it's being executed: 
    * if it's in a pre-installation context (such as anaconda when run as a `postscript`), it could use `ip link`, as it just needs to set the configuration files for the next boot, but doesn't really need to apply it yet, and the status of the interface doesn't really matter 
    * if it's in a post-installation context (such as a `postbootscript`) then applying configuration matters and it should use `nmcli`

**Comment by kcgthb** [comment]
Mmh, I may have to retract that: ~@bybai Yes, it takes a few tries, but it succeeds~
I think this was executed as a `postbootscript`, not as a `postscript`, so I can't confirm that the `ip link` lines properly work in a chrooted installation environment without proper testing and validation.

**Comment by bybai** [comment]
Hi @kcgthb, 

1. If it's in a pre-installation context (such as anaconda when run as a postscript), as it just needs to set the configuration files for the next boot, and the status of the interface doesn't really matter, and ``confignetwork`` also delete|add new temporary IPs using ``ip addr del dev ...|ip addr add dev ...``, ``confignetwork`` prepares the new ifcfg-* files, that is enough. So for diskful provison postscript stage, no need and will not execute ``ifdown/ifup`` to take the configure file effect. See PR #5430

2. If it is in ``postbootscript|updatenode`` stage, ``NetworkManager`` should be installed by default. So ``ifdown/ifup`` should work to take ifcfg-* files effect. 

3. In your environment, if ``ip link`` can work your problem, you can use it. Or else, you should check why ``NetworkManager`` is not installed, and install ``NetworkManager``. ``confignetwork`` will disable|stop ``NetworkManager``, but ``ifdown|ifup`` can use ``nmcli`` which is intalled by ``NetworkManager``.

**Comment by bybai** [comment]
verified by autotest environment.

**Comment by kcgthb** [comment]
Hi @bybai 
I can see an issue when this is executed in a pre-installation context (as a `postscript`) where `ip link` doesn't set up IP addresses, and other postscripts which require IPs (such as ``setroute`) would fail. See https://github.com/xcat2/xcat-core/pull/5430#issuecomment-409404709 for more details.

**Comment by kcgthb** [comment]
Hi @bybai. How has this been resolved? I don't see any update on #5430.

**Comment by bybai** [comment]
Hi @kcgthb, 
I am still working on this issue. 
1. why do you use `confignetwork` and `setroute` in postscript not in postbootscripts?

2. In redhat7.x environment, I can find `nmcli` in the kickstart/anaconda environment. So `ifup|ifdown` can work during `post-scripts` stage.  I am curious that there is no `nmcli` in Centos7, could you have a try to add `NetworkManager` in pkglist? 

3. As we discussed before, if using `ip link` instead of `ifup/ifdown` seems a much better approach: error messages about nmcli don't appear anymore. Could you have a try to use `ip link` in your own code ?  

4. In #5430, for diskful provision postscript stage, no need and will not execute "ifdown/ifup" to take the configure file effect, we also do not add "ip link" `up|down` for bond in `confignetwork`. We  can re-evaluate to add "ifdown|ifup" during diskful provision postscript stage, but now "no nmcli" in your env, and it can also not affect your case.

**Comment by kcgthb** [comment]
Hi @bybai 
1. why should I? :) I prefer doing going through the most possible tasks in the initial deployment phase, so nodes are ready to work as soon as possible after their first boot. I also have postscripts that depend on that network configuration and routes to be done before they can be executed.
2. that's actually a very good point, thanks for pointing that out! I reviewed our whole deployment process and it looks like NetworkManager was removed as a consequence of our ospks/otherpkgs lists. Adjusting those lists prevents NetworkManager to be removed and it can now be found by the `confignetwork` script.

Given 2, I guess that points 3 and 4 are moot now, and using `ifup/ifdown` should work out-of-the box in the deployment environment. The issue was with our specific setup.

**Comment by bybai** [comment]
Hi @kcgthb,
1. In diskful post-scripts stage, `confignetwork` and `setroute` only generate the necessary configuration files, invoke operations may cause failure, so should not invoke operations. 

2. I saw `setroute` failed in your case, could you try to use `setroute add`?  `setroute` default action is `replace`, if you use `add` operation in postscripts, although "Adding temporary route" failure, it can add persistent route in `/etc/sysconfig/static-routes`, which is different from `replace` operation.

3. If your postscripts only depend on that network|routes configuration files, that should work. If you need network|routes take effect in post-scripts stage, I think you should install `NetworkManager` and run old `confignetwork`(or latest `confignetwork`, `ifdown <interface>;ifup <interface>`), after that, removing `NetworkManager`, that should work in your case. This is your specific setup, you can realize these in your own postscripts.

**Comment by bybai** [comment]
Hi @kcgthb, Could you have a try based on my last comments?

**Comment by bybai** [comment]
@kcgthb, thanks for your quick reply. As it just needs to set the configuration files for the next boot, and the status of the interface doesn't really matter, we will keep current  `confignetwork` changes for diskful postscripts just keep the same action with `confignics`.
I will close this issue, if any needed, we can re-open it later or open a new issue.

**Comment by rodceron** [comment]
@kcgthb @bybai Good day! I'm curious as to how you've solved the problem of bringing the interfaces up and running after a RHEL 7 install and before first boot (ie, during postscripts). Like @kcgthb, I also needed networking to be up and  running during postscripts and run setroute replace  because my default gateway lies on an interface other than  the OS deployment one. I did that so that my target systems could connect to RHSM to register themselves and perform a yum update, mainly to get the latest and greatest kernel. Then, later in postbootscripts that new kernel would be up and  running already and I would be able to compile kernel-dependent modules (for instance: mellanox ofed drivers, gpfs, nvidia gpu drivers...). I tried to write a simple script that called ifup during postscripts but that just hangs around for a long time. So, my question is what are the exact steps to simulate the old behavior of confignetwork -s to not only create config  files, but also bring networking up in postscripts phase? Thank you!
